File: pkg/python/examples/mlx_example.py
#!/usr/bin/env python3
"""
Lux Consensus MLX GPU Acceleration Example

Demonstrates Apple Silicon GPU acceleration using MLX framework.
"""

import random
import time
from lux_consensus.mlx_backend import (
    MLXConsensusBackend,
    AdaptiveMLXBatchProcessor
)


def generate_vote():
    """Generate random vote for testing"""
    voter_id = bytes([random.randint(0, 255) for _ in range(32)])
    block_id = bytes([random.randint(0, 255) for _ in range(32)])
    is_preference = random.choice([True, False])
    return (voter_id, block_id, is_preference)


def main():
    print("=== Lux Consensus MLX GPU Acceleration Demo ===\n")

    # Initialize backend
    backend = MLXConsensusBackend(
        device_type="gpu",
        batch_size=32,
        enable_quantization=True,
        cache_size=5000
    )

    print(f"Device: {backend.get_device_name()}")
    print(f"GPU Enabled: {backend.gpu_enabled}\n")

    # Performance benchmarks
    print("Performance Benchmarks:")
    print("=======================\n")

    batch_sizes = [10, 100, 1000, 10000]

    for batch_size in batch_sizes:
        # Generate test votes
        votes = [generate_vote() for _ in range(batch_size)]

        # Warm-up
        backend.process_votes_batch(votes)

        # Benchmark
        start = time.perf_counter()
        processed = backend.process_votes_batch(votes)
        end = time.perf_counter()

        duration = (end - start) * 1_000_000  # microseconds
        throughput = batch_size / (end - start) if (end - start) > 0 else 0

        print(f"Batch Size: {batch_size}")
        print(f"  Time: {duration:.0f} μs")
        print(f"  Throughput: {throughput:.0f} votes/sec")
        print(f"  Per-vote: {duration/batch_size:.0f} ns")
        print(f"  Processed: {processed}/{batch_size}\n")

    # Memory usage
    print("GPU Memory Usage:")
    print(f"  Active: {backend.get_gpu_memory_usage() / (1024*1024):.1f} MB")
    print(f"  Peak: {backend.get_peak_gpu_memory() / (1024*1024):.1f} MB\n")

    # Test block validation
    print("Testing Block Validation:")
    print("=========================\n")

    block_ids = [bytes([random.randint(0, 255) for _ in range(32)]) for _ in range(100)]

    start = time.perf_counter()
    results = backend.validate_blocks_batch(block_ids)
    end = time.perf_counter()

    duration = (end - start) * 1_000_000
    print(f"Validated 100 blocks in {duration:.0f} μs")
    print(f"Valid blocks: {sum(results)}/100\n")

    # Test adaptive batch processor
    print("Testing Adaptive Batch Processor:")
    print("==================================\n")

    processor = AdaptiveMLXBatchProcessor(backend)

    start = time.perf_counter()

    # Process 10,000 votes with adaptive batching
    for _ in range(10000):
        vote = generate_vote()
        processor.add_vote(*vote)

    processor.flush()
    end = time.perf_counter()

    duration = (end - start) * 1_000_000
    print(f"Total time: {duration:.0f} μs")
    print(f"Throughput: {processor.get_throughput():.0f} votes/sec")
    print(f"Optimal batch size: {processor.get_batch_size()}\n")

    # CPU vs GPU comparison
    print("CPU vs GPU Comparison:")
    print("======================\n")

    # CPU mode
    cpu_backend = MLXConsensusBackend(device_type="cpu", batch_size=100)
    votes = [generate_vote() for _ in range(1000)]

    start = time.perf_counter()
    cpu_backend.process_votes_batch(votes)
    end = time.perf_counter()
    cpu_time = (end - start) * 1_000_000

    # GPU mode
    gpu_backend = MLXConsensusBackend(device_type="gpu", batch_size=100)

    start = time.perf_counter()
    gpu_backend.process_votes_batch(votes)
    end = time.perf_counter()
    gpu_time = (end - start) * 1_000_000

    speedup = cpu_time / gpu_time if gpu_time > 0 else 0

    print(f"CPU Time: {cpu_time:.0f} μs")
    print(f"GPU Time: {gpu_time:.0f} μs")
    print(f"Speedup: {speedup:.1f}x\n")

    print("✅ MLX GPU acceleration working!")


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"Error: {e}")
        print("\nMake sure MLX is installed: pip3 install mlx")
        exit(1)

File: pkg/cpp/examples/mlx_example.cpp
#include "lux/consensus.hpp"

#ifdef HAS_MLX
#include "lux/mlx_consensus.hpp"
#endif

#include <chrono>
#include <iostream>
#include <random>
#include <vector>

using namespace lux::consensus;

int main() {
#ifdef HAS_MLX
    std::cout << "=== Lux Consensus MLX GPU Acceleration Demo ===\n\n";

    // Configure MLX
    MLXConsensus::MLXConfig mlx_config{
        .model_path = "/models/consensus/mlx_model.bin",
        .device_type = "gpu",
        .batch_size = 32,
        .enable_quantization = true,
        .cache_size = 5000,
        .parallel_ops = 8
    };

    try {
        // Create MLX consensus engine
        auto mlx = std::make_unique<MLXConsensus>(mlx_config);

        std::cout << "Device: " << mlx->get_device_name() << "\n";
        std::cout << "GPU Enabled: " << (mlx->is_gpu_enabled() ? "Yes" : "No") << "\n\n";

        // Generate test votes
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<uint8_t> dist(0, 255);

        auto generate_vote = [&]() -> Vote {
            Vote vote;
            for (auto& byte : vote.voter_id.data) byte = dist(gen);
            for (auto& byte : vote.block_id.data) byte = dist(gen);
            vote.is_preference = (dist(gen) % 2) == 0;
            return vote;
        };

        // Benchmark different batch sizes
        std::vector<size_t> batch_sizes = {10, 100, 1000, 10000};

        std::cout << "Performance Benchmarks:\n";
        std::cout << "=======================\n\n";

        for (auto batch_size : batch_sizes) {
            std::vector<Vote> votes;
            votes.reserve(batch_size);
            for (size_t i = 0; i < batch_size; ++i) {
                votes.push_back(generate_vote());
            }

            // Warm-up
            mlx->process_votes_batch(votes);

            // Benchmark
            auto start = std::chrono::high_resolution_clock::now();
            auto processed = mlx->process_votes_batch(votes);
            auto end = std::chrono::high_resolution_clock::now();

            auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
            double throughput = batch_size * 1000000.0 / duration.count();

            std::cout << "Batch Size: " << batch_size << "\n";
            std::cout << "  Time: " << duration.count() << " μs\n";
            std::cout << "  Throughput: " << static_cast<size_t>(throughput) << " votes/sec\n";
            std::cout << "  Per-vote: " << (duration.count() / batch_size) << " ns\n";
            std::cout << "  Processed: " << processed << "/" << batch_size << "\n\n";
        }

        // Memory usage
        std::cout << "GPU Memory Usage:\n";
        std::cout << "  Active: " << (mlx->get_gpu_memory_usage() / (1024 * 1024)) << " MB\n";
        std::cout << "  Peak: " << (mlx->get_peak_gpu_memory() / (1024 * 1024)) << " MB\n\n";

        // Test adaptive batch processor
        std::cout << "Testing Adaptive Batch Processor:\n";
        std::cout << "==================================\n\n";

        AdaptiveMLXBatchProcessor processor(std::move(mlx));

        auto start = std::chrono::high_resolution_clock::now();

        // Process 10,000 votes with adaptive batching
        for (size_t i = 0; i < 10000; ++i) {
            processor.add_vote(generate_vote());
        }
        processor.flush();

        auto end = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);

        std::cout << "Total time: " << duration.count() << " μs\n";
        std::cout << "Throughput: " << static_cast<size_t>(processor.get_throughput()) << " votes/sec\n";
        std::cout << "Optimal batch size: " << processor.get_batch_size() << "\n\n";

        std::cout << "✅ MLX GPU acceleration working!\n";

        return 0;

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }

#else
    std::cout << "MLX support not enabled. Build with -DHAS_MLX=ON\n";
    return 1;
#endif
}
